// xgplayer/src/proxy.js
import EventEmitter from 'event-emitter'
import util from './utils/util'
import Errors from './error'
const ua = navigator.userAgent
const IS_SAFARI = ua.indexOf('Safari') > -1 && ua.indexOf('mac') > -1
// import allOff from 'event-emitter/all-off'
class Proxy {
  constructor (options) {
    this._hasStart = false
    this._seekingTimer = null
    this.videoConfig = {
      controls: false,
      autoplay: options.autoplay,
      playsinline: options.playsinline,
      'webkit-playsinline': options.playsinline,
      'x5-playsinline': options.playsinline,
      'x5-video-player-type': options['x5-video-player-type'],
      'x5-video-player-fullscreen': options['x5-video-player-fullscreen'],
      'x5-video-orientation': options['x5-video-orientation'],
      airplay: options['airplay'],
      'webkit-airplay': options['airplay'],
      tabindex: 2,
      mediaType: options.mediaType || 'video'
    }
    if (options.loop) {
      this.videoConfig.loop = 'loop'
    }
    if (options.crossorigin) {
      this.videoConfig.crossorigin = 'anonymous'
    }
    let textTrackDom = ''
    if (options.textTrack && Array.isArray(options.textTrack)) {
      options.textTrack.some(track => {
        if (track.src && track.label && track.default) {
          textTrackDom += `<track src="${track.src}" `
          if (track.kind) {
            textTrackDom += `kind="${track.kind}" `
          }
          textTrackDom += `label="${track.label}" `
          if (track.srclang) {
            textTrackDom += `srclang="${track.srclang}" `
          }
          textTrackDom += `${track.default ? 'default' : ''}>`
          return true
        }
      })
      this.videoConfig.crossorigin = 'anonymous'
    }
    if (options.textTrackStyle) {
      let style = document.createElement('style')
      this.textTrackStyle = style
      document.head.appendChild(style)
      let styleStr = ''
      for (let index in options.textTrackStyle) {
        styleStr += `${index}: ${options.textTrackStyle[index]};`
      }
      let wrap = options.id ? `#${options.id}` : (options.el.id ? `#${options.el.id}` : `.${options.el.className}`)
      if (style.sheet.insertRule) {
        style.sheet.insertRule(`${wrap} video::cue { ${styleStr} }`, 0)
      } else if (style.sheet.addRule) {
        style.sheet.addRule(`${wrap} video::cue`, styleStr)
      }
    }
    let el = options.el ? options.el : util.findDom(document, `#${options.id}`)
    const XgVideoProxy = this.constructor.XgVideoProxy
    if(XgVideoProxy && this.videoConfig.mediaType === XgVideoProxy.mediaType) {
      this.video = new XgVideoProxy(el, options)
    } else {
      this.video = util.createDom(this.videoConfig.mediaType, textTrackDom, this.videoConfig, '')
    }
    if (options.autoplay) {
      // this.video.autoplay = true
      if (options.autoplayMuted) {
        this.video.muted = true
      }
    }
    this.ev = ['play', 'playing', 'pause', 'ended', 'error', 'seeking', 'seeked',
      'timeupdate', 'waiting', 'canplay', 'canplaythrough', 'durationchange', 'volumechange', 'loadeddata', 'loadedmetadata', 'loadstart'
    ].map((item) => {
      return {
        [item]: `on${item.charAt(0).toUpperCase()}${item.slice(1)}`
      }
    })
    EventEmitter(this)

    this._interval = {}
    let lastBuffer = '0,0'
    let self = this

    this.ev.forEach(item => {
      self.evItem = Object.keys(item)[0]
      let name = Object.keys(item)[0]
      self.video.addEventListener(Object.keys(item)[0], function () {
        // fix when video destroy called and video reload
        if (!self.src) {
          return
        }
        if (name === 'play') {
          self.hasStart = true
        } else if (name === 'waiting') {
          self.inWaitingStart = new Date().getTime()
        } else if (name === 'playing') {
          if (self.inWaitingStart) {
            self.inWaitingStart = undefined
          }
        }
        if (name === 'error') {
          self._onError()
        } else {
          self.emit(name, self)
        }

        if (self.hasOwnProperty('_interval')) {
          if (['ended', 'error', 'timeupdate'].indexOf(name) < 0) {
            clearInterval(self._interval['bufferedChange'])
            util.setInterval(self, 'bufferedChange', function () {
              let curBuffer = []
              for (let i = 0, len = self.video.buffered.length; i < len; i++) {
                curBuffer.push([self.video.buffered.start(i), self.video.buffered.end(i)])
              }
              if (curBuffer.toString() !== lastBuffer) {
                lastBuffer = curBuffer.toString()
                self.emit('bufferedChange', curBuffer)
              }
            }, 200)
          } else {
            if (name !== 'timeupdate') {
              util.clearInterval(self, 'bufferedChange')
            }
          }
        }
      }, false)
    })
  }

  _onError () {
    const self = this
    if (self.video.error) {
      if (self.video.error.message && self.video.error.message.indexOf('Empty src attribute') > -1) {
        return
      }
      self.emit('error', new Errors('other', self.currentTime, self.duration, self.networkState, self.readyState, self.currentSrc, self.src,
        self.ended, {
          line: 41,
          msg: self.error,
          handle: 'Constructor',
          code: self.errorCode
        }))
    }
  }

  get hasStart () {
    return this._hasStart
  }
  set hasStart (bool) {
    if (typeof bool === 'boolean' && bool === true && !this._hasStart) {
      this._hasStart = true
      this.emit('hasstart')
    }
  }
  destroy () {
    // allOff(this)
    if (this.textTrackStyle) {
      this.textTrackStyle.parentNode.removeChild(this.textTrackStyle)
    }
  }
  play () {
    return this.video.play()
  }
  pause () {
    this.video.pause()
  }
  canPlayType () {
    this.video.canPlayType()
  }
  getBufferedRange (buffered) {
    let range = [0, 0]
    let video = this.video
    if(!buffered) buffered = video.buffered
    let currentTime = video.currentTime
    if (buffered) {
      for (let i = 0, len = buffered.length; i < len; i++) {
        range[0] = buffered.start(i)
        range[1] = buffered.end(i)
        if (range[0] <= currentTime && currentTime <= range[1]) {
          break
        }
      }
    }
    if (range[0] - currentTime <= 0 && currentTime - range[1] <= 0) {
      return range
    } else {
      return [0, 0]
    }
  }
  set autoplay (isTrue) {
    if (!this.video) return
    this.video.autoplay = isTrue
  }
  get autoplay () {
    if (!this.video) return null
    return this.video.autoplay
  }
  get buffered () {
    if (!this.video) return null
    return this.video.buffered
  }
  get buffered2 () {
    return util.getBuffered2(this.video.buffered)
  }
  get crossOrigin () {
    if (!this.video) return null
    return this.video.crossOrigin
  }
  set crossOrigin (isTrue) {
    if (!this.video) return
    this.video.crossOrigin = isTrue
  }
  get currentSrc () {
    if (!this.video) return null
    return this.video.currentSrc
  }
  set currentSrc (src) {
    if (!this.video) return
    this.video.currentSrc = src
  }
  get currentTime () {
    if (!this.video) return null
    return this.video.currentTime
  }
  set currentTime (time) {
    if (typeof isFinite === 'function' && !isFinite(time)) return
    if (this.video) {
      this.video.currentTime = time
      this.emit('currentTimeChange')
    }
  }
  get defaultMuted () {
    if (!this.video) return null
    return this.video.defaultMuted
  }
  set defaultMuted (isTrue) {
    if (!this.video) return
    this.video.defaultMuted = isTrue
  }
  get duration () {
    if(this.config.duration) {
      if(this.video) return Math.min(this.config.duration, this.video.duration)
      else return this.config.duration
    }
    if (!this.video) return null
    return this.video.duration
  }
  get ended () {
    if (!this.video) return null
    return this.video.ended
  }
  get error () {
    if (!this.video) return null
    let err = this.video.error
    if (!err) {
      return null
    }
    let status = [{
      en: 'MEDIA_ERR_ABORTED',
      cn: '取回过程被用户中止'
    }, {
      en: 'MEDIA_ERR_NETWORK',
      cn: '当下载时发生错误'
    }, {
      en: 'MEDIA_ERR_DECODE',
      cn: '当解码时发生错误'
    }, {
      en: 'MEDIA_ERR_SRC_NOT_SUPPORTED',
      cn: '不支持音频/视频'
    }]
    return this.lang ? this.lang[status[err.code - 1].en] : status[err.code - 1].en
  }
  get errorCode () {
    if (!this.video) return null
    let err = this.video.error || {}
    return err.code
  }
  get loop () {
    if (!this.video) return null
    return this.video.loop
  }
  set loop (isTrue) {
    if (!this.video) return
    this.video.loop = isTrue
  }
  get muted () {
    if (!this.video) return null
    return this.video.muted
  }
  set muted (isTrue) {
    if (!this.video) return
    this.video.muted = isTrue
  }
  get networkState () {
    if (!this.video) return null
    let status = [{
      en: 'NETWORK_EMPTY',
      cn: '音频/视频尚未初始化'
    }, {
      en: 'NETWORK_IDLE',
      cn: '音频/视频是活动的且已选取资源，但并未使用网络'
    }, {
      en: 'NETWORK_LOADING',
      cn: '浏览器正在下载数据'
    }, {
      en: 'NETWORK_NO_SOURCE',
      cn: '未找到音频/视频来源'
    }]
    return this.lang ? this.lang[status[this.video.networkState].en] : status[this.video.networkState].en
  }
  get paused () {
    if (!this.video) return null
    return this.video.paused
  }
  get playbackRate () {
    if (!this.video) return null
    return this.video.playbackRate
  }
  set playbackRate (rate) {
    if (!this.video) return
    this.video.playbackRate = rate
  }
  get played () {
    if (!this.video) return null
    return this.video.played
  }
  get preload () {
    if (!this.video) return null
    return this.video.preload
  }
  set preload (isTrue) {
    if (!this.video) return
    this.video.preload = isTrue
  }
  get readyState () {
    if (!this.video) return null
    let status = [{
      en: 'HAVE_NOTHING',
      cn: '没有关于音频/视频是否就绪的信息'
    }, {
      en: 'HAVE_METADATA',
      cn: '关于音频/视频就绪的元数据'
    }, {
      en: 'HAVE_CURRENT_DATA',
      cn: '关于当前播放位置的数据是可用的，但没有足够的数据来播放下一帧/毫秒'
    }, {
      en: 'HAVE_FUTURE_DATA',
      cn: '当前及至少下一帧的数据是可用的'
    }, {
      en: 'HAVE_ENOUGH_DATA',
      cn: '可用数据足以开始播放'
    }]
    return this.lang ? this.lang[status[this.video.readyState].en] : status[this.video.readyState]
  }
  get seekable () {
    if (!this.video) return null
    return this.video.seekable
  }
  get seeking () {
    if (!this.video) return null
    return this.video.seeking
  }
  get src () {
    if (!this.video || !this.video.src || this.video.src === location.href) return null
    return this.video.src
  }
  set src (url) {
    if (!this.video) return
    this.video.pause()
    if(typeof this.config.beforeStartRun === 'function') {
      this.config.beforeStartRun().then(() => {
        this.video.src = url
        this.emit('srcChange')
      })
    } else {
      this.video.src = url
      this.emit('srcChange')
    }
  }
  get volume () {
    if (!this.video) return null
    return this.video.volume
  }
  set volume (vol) {
    if (typeof isFinite === 'function' && !isFinite(vol)) return
    if (!this.video) return
    this.video.volume = vol
  }
  get fullscreen () {
    return util.hasClass(this.root, 'xgplayer-is-fullscreen') || util.hasClass(this.root, 'xgplayer-fullscreen-active')
  }
  get bullet () {
    return util.findDom(this.root, 'xg-bullet') ? util.hasClass(util.findDom(this.root, 'xg-bullet'), 'xgplayer-has-bullet') : false
  }
  get textTrack () {
    return util.hasClass(this.root, 'xgplayer-is-textTrack')
  }
  get pip () {
    return util.hasClass(this.root, 'xgplayer-pip-active')
  }
}

export default Proxy
